### DISCUSSION TITLE
Moving the ," discussion to discussion

### DISCUSSION THREAD

**Discussion by tabemann** [tabemann/zeptoforth#40] [General]
@ctrilsbach @robzed

Here is a simple implementation of `,"`:

```
: ," advance-once [char] " internal::parse-to-char 255 min dup c, over + swap ?do i c@ c, loop ;
```

Note that this implementation is only meant for interpretation mode; if invoked from compiled code it will search the parse buffer for " when the compiled code is executed and will only then append it to the current dictionary. Also note that this appends a counted string including the initial count byte.

Here is an example of it in action:

```
create foo ," FOOBAR!"  ok
foo dup 8 + dump 
20015EE4  07 46 4F 4F  42 41 52 21  -- -- -- --  -- -- -- --  |.FOOBAR!........|
 ok
```

**Comment by tabemann** [comment]
I must admit that I never implemented `word` because for most purposes it is not useful to stop parsing at a single character; the idiomatic word to use for parsing a single token is `token` which is different from standard `$20 word` in that it will stop at any kind of whitespace and not just a single character. I also just discovered that I never implemented standard `bl` ─ it will now be in the next release.

**Comment by robzed** [comment]
Great definition. 
One question - why did you use a loop with c, vs. cstring, ?
Is there something I'm missing with cstring,  ??
To be honest, I'm not sure what internal::parse-to-char returns (I should check.

**Comment by tabemann** [comment]
@robzed You know, I pretty much forgot about `cstring,` ─ here is the same, but now with `cstring,`, which makes for much more succinct code. This is even though zeptoforth use `cstring,` internally in just about any place a string literal needs to be compiled.

As for `internal::parse-to-char`, that returns an address and a length of the parsed string, which is valid only as long as the contents of the evaluation buffer are valid.

```
: ," advance-once [char] " internal::parse-to-char cstring, ;  ok
create foo ," FOOBAR!"  ok
foo dup 8 + dump 
20015E88  07 46 4F 4F  42 41 52 21  -- -- -- --  -- -- -- --  |.FOOBAR!........|
 ok
```

**Reply by robzed** [reply]
That is amazingly clear, compared to the previous version. 

It reminds me of the advice I give people who complain about stack juggling; refactor into separate words. I have to remember this myself... and it helps with testing as well.

**Reply by robzed** [reply]
I also realize what you are doing with the do-loop now. I'd have done it with a begin-while-repeat look I think, although the cstring version is so amazingly clear I love it!

**Reply by robzed** [reply]
@ctrilsbach this last version of ," is the one I'd use with Zeptoforth. It's elegant and will hopefully do what you want. 

I think this will perform as per Forth application techniques. The word is marked with a stylized 4 to indicate SwiftForth specific - but usually it's pretty easy to make your own, and good practice. 

We are also fortunate to have @tabemann so proactive with his Zeptoforth.

**Reply by ctrilsbach** [reply]
@tabemann @robzed Thanks to you both. As said in issues, C" worked fine so I hope you haven't spent a lot of unecessary time on this.
